### PR TITLE
Pract.02 : Add 04-07 | Gabriela Mladenova

### PR DESCRIPTION
**Pull request by gabrielavm**
Gabriela Mladenova
SI 1 3MI0600225

### CODE REVIEW

File: Seminars/Sem.02/Pract.02/Task04.cpp [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+Relation initRelation()
+{
+    Relation newRelation;
+
+    newRelation.numberOfPairs = 0;
+
+    return newRelation;
+}
+
+void addPairToRelation(Relation& newRelation, Pair& newPair)
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task04.cpp]: const Pair&

File: Seminars/Sem.02/Pract.02/Task04.cpp [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    for (int i = 0; i < newRelation.numberOfPairs; ++i)
+    {
+        writeToFile << newRelation.pairs[i].firstNumber << " , "
+            << newRelation.pairs[i].secondNumber << std::endl;
+    }
+
+    writeToFile.close();
+}
+
+Relation readPairFromFile()
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task04.cpp]: this name is ambiguos
You read a relation from the file in this funciton

File: Seminars/Sem.02/Pract.02/Task07.cpp [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+//SI
+//Group 1
+//FN: 3MI0600225
+
+#include <iostream>
+#include <fstream>
+
+const int SIZE = 128;
+const int USERS = 100;
+const char FILENAME1[] = "usersList.txt";
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task07.cpp]: FILENAME1 is not good name

File: Seminars/Sem.02/Pract.02/Task07.cpp [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    }
+
+    int i;
+    for (i = 0; i < USERS; ++i)
+    {
+        readFromFile >> newSystem.systemOfUsers[i].name
+            >> newSystem.systemOfUsers[i].email
+            >> newSystem.systemOfUsers[i].password;
+    }
+    newSystem.countOfStudentsInSystem = i;
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task07.cpp]: The idea is to store the current number of users in countOfStudentsInSystem, not the maximum possible size

File: Seminars/Sem.02/Pract.02/Task07.cpp [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    std::cin >> newSystem.systemOfUsers[i].email;
+    ++i;
+
+    std::cout << "Enter your password: ";
+    std::cin >> newSystem.systemOfUsers[i].password;
+    i = 0;
+
+    inputUsers(newSystem);
+
+    std::cout << "You are successfully registered!" << std::endl;
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task07.cpp]: What is the logic here???

File: Seminars/Sem.02/Pract.02/Task07.cpp [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+int strCompare(char str1[], char str2[])
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task07.cpp]: use const

File: Seminars/Sem.02/Pract.02/Task07.cpp [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    {
+        if (str1[i] == str2[i])
+        {
+            continue;
+        }
+        else
+        {
+            break;
+        }
+    }
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task07.cpp]: just if(str[i]!=str2[i]) break;

File: Seminars/Sem.02/Pract.02/Task07.cpp [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+    if (str1[i] == str2[i])
+    {
+        return 1;
+    }
+
+    return 0;
+}
+
+void logInTheSystem(System& newSystem)
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task07.cpp]: const

**Review by khatoni — CHANGES_REQUESTED** [review]

**Comment by Justsvetoslavov** [comment]
@khatoni Review the PR again.
Specifically interested in Task07 as the current one has flaws.

File: Seminars/Sem.02/Pract.02/Task07.cpp [hunk 9]
```diff
@@ -46,55 +46,49 @@ void inputUsers(System &newSystem)
...
 
-    int i;
-    for (i = 0; i < USERS; ++i)
+    int count = 0;
+    for (int i = 0; i < USERS; ++i)
     {
         readFromFile >> newSystem.systemOfUsers[i].name
             >> newSystem.systemOfUsers[i].email
             >> newSystem.systemOfUsers[i].password;
+
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task07.cpp]: Again why you want to read the maximum number of students. What will happen when you try to read from the file the maximum possible size of students and the file has less users saved in it.

File: Seminars/Sem.02/Pract.02/Task07.cpp [hunk 10]
```diff
@@ -25,7 +25,7 @@ struct System
 
 void inputUsers(System &newSystem)
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task07.cpp:26]: I do not understand the logic here. As I can see, you want to save the whole system to file, therefore you do not change it and you should use const . Why do you change the countOfStudents . Re-think the logic here. Why the for cycle is looping to the maximum possible number of users. It should be to the current number of users in the system

File: Seminars/Sem.02/Pract.02/Task07.cpp [hunk 11]
```diff
@@ -46,55 +46,49 @@ void inputUsers(System &newSystem)
...
     readFromFile.close();
 }
 
 void makeRegistration(System& newSystem)
 {
-    int i = 0;
     std::cout << "Enter your name: ";
-    std::cin >> newSystem.systemOfUsers[i].name;
-    ++i;
+    std::cin >> newSystem.systemOfUsers[0].name;
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task07.cpp:73]: Why do you store the name of the new user at index 0

File: Seminars/Sem.02/Pract.02/Task07.cpp [hunk 12]
```diff
@@ -46,55 +46,49 @@ void inputUsers(System &newSystem)
...
 
     std::cout << "Enter your email: ";
-    std::cin >> newSystem.systemOfUsers[i].email;
-    ++i;
+    std::cin >> newSystem.systemOfUsers[1].email;
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task07.cpp:76]: Why do you store the email of the new user at index 1, which is different index from index 0

File: Seminars/Sem.02/Pract.02/Task07.cpp [hunk 13]
```diff
@@ -108,7 +102,7 @@ int strCompare(char str1[], char str2[])
     return 0;
```
Comment (khatoni) [inline Seminars/Sem.02/Pract.02/Task07.cpp:102]: You can just write:  return str1[i]==str2[i]

**Review by khatoni — CHANGES_REQUESTED** [review]
I think that your idea behind the functions printUsersList ,inputUsers and makeRegistration is not clear enough. Please chech the uploaded solution for this task.

**Pull request closed without merging**